url8d3c4d17f5_catseye_JaC64_tgz-pJ8-JaC64J8/src/com/dreamfabric/jac64/CIA.java

```diff
@@ -0,0 +1,617 @@
+        // Since this should continue run, just reschedule...
+//        System.out.println("TOD ticked..." + (todhour>>4) + (todhour & 0xf) + ":" + 
+//              (todmin>>4) + (todmin&0xf) + ":" + (todsec>>4) + (todsec&0xf));
+        cpu.scheduler.addEvent(this);
+      }
+  };
+  
+  /**
+   * Creates a new <code>CIA</code> instance.
+   *
+   */
+  public CIA(MOS6510Core cpu, int offset, ExtChip chips) {
+    this.cpu = cpu;
+    this.offset = offset;
+    this.chips = chips;
+    timerA = new CIATimer("TimerA", 1, true, null);
+    timerB = new CIATimer("TimerB", 2, false, timerA);
+    timerA.otherTimer = timerB;
+    todEvent.time = cpu.cycles + 10000;
+    cpu.scheduler.addEvent(todEvent);
+  }
+
+  public void reset() {
+    ciaicrRead = 0;
+    ciaie = 0;
+
+    timerA.reset();
+    timerB.reset();
+    tod10sec = 0;
+    todsec = 0;
+    todmin = 0;
+    todhour = 0;
+
+    updateInterrupts();
+  }
+
+  public String ciaID() {
+    return offset == 0x10c00 ? "CIA 1" : "CIA 2";
+  }
+
+  public int performRead(int address, long cycles) {
+    address -= offset;
+    switch(address) {
+    case DDRA:
+      return ddra;
+    case DDRB:
+      return ddrb;
+    case PRA:
+      return (pra | ~ddra) & 0xff;
+    case PRB:
+      int data = (prb | ~ddrb) & 0xff;
+      if ((timerA.cr & 0x02) > 0) {
+        data &= 0xbf;
+        if ((timerA.cr & 0x04) > 0) {
+          data |= timerA.flipflop ? 0x40 : 0;
+        } else {
+          data |= timerA.underflow ? 0x40 : 0;          
+        }
+      }
+      if ((timerB.cr & 0x02) > 0) {
+        data &= 0x7f;
+        if ((timerB.cr & 0x04) > 0) {
+          data |= timerB.flipflop ? 0x80 : 0;
+        } else {
+          data |= timerB.underflow ? 0x80 : 0;          
+        }
+      }
+      return data;
+    case TIMALO:
+//      System.out.println(ciaID() + " getTimerA LO, timer = " + timerA.getTimer(cycles));
+      return timerA.getTimer(cycles) & 0xff;
+    case TIMAHI:
+      return timerA.getTimer(cycles) >> 8;
+    case TIMBLO:
+//      System.out.println(ciaID() + " getTimerB LO, timer = " + timerB.getTimer(cycles));
+      return timerB.getTimer(cycles) & 0xff;
+    case TIMBHI:
+      return timerB.getTimer(cycles) & 0xff;
+    case TODTEN:
+      return tod10sec;
+    case TODSEC:
+      return todsec;
+    case TODMIN:
+      return todmin;
+    case TODHRS:
+      return todhour;
+    case SDR:
+      return sdr;
+    case CRA:
+      return timerA.cr;
+    case CRB:
+      return timerB.cr;
+    case ICR:
+      // Clear interrupt register (flags)!
+      if (TIMER_DEBUG && offset == 0x10d00) println("clear Interrupt register, was: " +
+          Hex.hex2(ciaicrRead));
+      int val = ciaicrRead;
+      ciaicrRead = 0;
+
+      // Latch off the IRQ/NMI immediately!!!
+      updateInterrupts();
+
+      return val;
+      default:
+        return 0xff;
+    }
+  }
+
+  public void performWrite(int address, int data, long cycles) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 63 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.